feat: Refactor ConversionGoalFilter schema and related components

jabahamondes opened the pull request:
## Changes

This PR includes:
- migrating logic from frontend to backend query runner
- conversion goal support for action and more filters
- marketing table support for sorting and loading state

Note: there is a lot of improvements that can be done in the future to make this better. Some of those improvement are:
- using ast expression for queries
- removing duplicated logic in the frontend/backend
- TBD more


<img width="984" alt="image" src="https://github.com/user-attachments/assets/eef99458-b001-4f03-91ca-f66fb7aa2344" />
<img width="929" alt="image" src="https://github.com/user-attachments/assets/02b1d80d-0fb0-424e-a428-8256cf8d86b7" />
<img width="1219" alt="image" src="https://github.com/user-attachments/assets/149e6964-2585-4328-9522-4acd7bb8420b" />
<img width="1205" alt="image" src="https://github.com/user-attachments/assets/6426ac13-8ac9-44a9-96cf-9350cec81950" />

Commit:
feat: Refactor ConversionGoalFilter schema and related components

- Removed deprecated ActionFilter schema from schema.json.
- Updated ConversionGoalFilter to support new properties and structure, including event handling and fixed properties.
- Adjusted ConversionGoalDropdown and ConversionGoalsConfiguration components to align with the new filter structure.
- Enhanced validation logic for conversion goals in the backend to accommodate new NodeKind types.
- Cleaned up unused imports and code in various files for better maintainability.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

jabahamondes committed:
Working first basic (#34116)

Co-authored-by: Javier Bahamondes <javierbahamondes@Javiers-MacBook-Pro.local>

jabahamondes commented on products/marketing_analytics/backend/hogql_queries/test/bigquery_data/table1.csv:
```diff
@@ -0,0 +1,26 @@
+date1,spend1,clicks1,keyworks1,impressions1,campaign1
```
These are table I created for future testing

robbie-c commented:
For this UI 
<img width="343" alt="Screenshot 2025-06-26 at 11 13 46" src="https://github.com/user-attachments/assets/ebe81613-5be0-4edb-928e-7fb46666a769" />

Could we let people leave the name blank, and in that case use the event/action/etc name?

robbie-c reviewed (changes requested):
Left some comments.

Some are comments on things that technically weren't changed in this PR, feel free to deal with those in a different PR

It'd be amazing if generate_demo_data could set up some fake data for me to click around in - out of scope for this PR but would be a really nice thing to do long-term

Commit:
Merge conflicts fix

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

Commit:
Prs comments

Commit:
Merge branch 'feat/conversion-goal-v2-ma' of github.com:PostHog/posthog into feat/conversion-goal-v2-ma

github-actions[bot] committed:
Update UI snapshots for `chromium` (13)

github-actions[bot] committed:
Update query snapshots

jabahamondes commented on posthog/hogql_queries/insights/trends/aggregation_operations.py:
```diff
@@ -173,18 +173,12 @@ def _math_func(self, method: str, override_chain: Optional[list[str | int]] = No
-                        ast.Call(name="toDateTime", args=[ast.Constant(value=0), ast.Constant(value="UTC")]),
-                    ],
-                )
+                timestamp_expr = ast.Call(name="today", args=[])
```
No. So the problem is that self managed sources don't have `date` types only string, so there was an error by converting the currency with "string" dates. Even if I casted to Datetime it encounter issues when it calls `parseDateTime64BestEffortOrNull` for the date

Commit:
Sanitize fn in utils

Commit:
Refactor marketing analytics query generation to use AST internally

- Updated , , , and other methods to utilize AST for building SQL queries.
- Introduced new methods for generating campaign cost CTE and base columns using AST.
- Adapted marketing source adapters (BigQuery, Google Ads, Self-Managed) to return AST expressions for field retrieval.
- Enhanced date range condition handling with AST for better query construction.

This change improves the maintainability and readability of the query generation logic.

Commit:
More Hogql migration code

Commit:
Add MarketingAnalyticsFilters component and integrate into WebAnalyticsDashboard

- Introduced MarketingAnalyticsFilters component for marketing-specific date filtering.
- Updated WebAnalyticsDashboard to conditionally render MarketingAnalyticsFilters based on the selected product tab.
- Modified WebAnalyticsHealthCheck to return null for the marketing tab, preventing unnecessary checks.
- Refactored conversion goal processor to enhance query generation with AST, improving maintainability.

This update enhances the user experience for marketing analytics by providing tailored filtering options.

Commit:
Refactor marketing analytics query generation and improve AST usage

- Updated  to streamline UTM expression retrieval and condition handling.
- Simplified query generation methods to return AST directly, enhancing maintainability.
- Refactored marketing source adapters to align with new method signatures for field retrieval.
- Improved date range condition handling across adapters for better query construction.

This update enhances the overall structure and readability of the marketing analytics query generation logic.

Commit:
Merge branch 'feat/conversion-goal-v2-ma' of github.com:PostHog/posthog into feat/conversion-goal-v2-ma

jabahamondes commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/frontend/components/MarketingAnalyticsTable.tsx:
```diff
@@ -0,0 +1,182 @@
+            // Add sortable column headers for marketing analytics fields
+            // These match the backend output column names
+            Campaign: {
+                renderTitle: MarketingSortableCell('Campaign', `${CAMPAIGN_COST_CTE_NAME}.${CAMPAIGN_NAME_FIELD}`),
```
I'll check this in a following PR if that's ok ✅

robbie-c replied:
Ok

jabahamondes commented on products/marketing_analytics/backend/hogql_queries/adapters/google_ads.py:
```diff
@@ -0,0 +1,134 @@
+            self.logger.exception("Google Ads validation failed", error=error_msg)
+            return ValidationResult(is_valid=False, errors=[error_msg])
+
+    def _get_campaign_name_field(self) -> str:
```
I change most of the places, do you think it's ok as is and keep improving it in the future?

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

Commit:
Fix formatting issues in MarketingAnalyticsFilters and WebAnalyticsDashboard; refactor conversion goal processor method names for clarity

- Added missing newline at the end of MarketingAnalyticsFilters.tsx.
- Corrected the closing brace in WebAnalyticsDashboard.tsx for proper syntax.
- Renamed method in ConversionGoalProcessor from  to  to better reflect its functionality and updated related calls.

These changes improve code readability and maintainability across the marketing analytics components.

Commit:
Merge remote-tracking branch 'origin/master' into feat/conversion-goal-v2-ma

Commit:
Merge branch 'origin/feat/conversion-goal-v2-ma' into feat/conversion-goal-v2-ma

jabahamondes commented on frontend/src/scenes/web-analytics/tabs/marketing-analytics/utils.ts:
```diff
@@ -23,6 +25,8 @@ export type MarketingAnalyticsSchema = keyof typeof MARKETING_ANALYTICS_SCHEMA
 export const MARKETING_ANALYTICS_CONVERSION_GOAL_SCHEMA: Record<string, MarketingAnalyticsSchemaField> = {
     [UTM_CAMPAIGN_NAME_SCHEMA_FIELD]: { type: ['string'], required: true },
     [UTM_SOURCE_NAME_SCHEMA_FIELD]: { type: ['string'], required: true },
```
👌  this will change in the future though

jabahamondes commented on products/marketing_analytics/backend/hogql_queries/adapters/bigquery.py:
```diff
@@ -0,0 +1,221 @@
+        if currency_field and total_cost_field:
+            # toFloat(convertCurrency('currency_field', 'base_currency', toFloat(coalesce(total_cost_field, 0))))
+            coalesce = ast.Call(name="coalesce", args=[ast.Field(chain=[total_cost_field]), ast.Constant(value=0)])
+            inner_toFloat = ast.Call(name="toFloat", args=[coalesce])
+            convert_currency = ast.Call(
+                name="convertCurrency",
+                args=[ast.Constant(value=currency_field), ast.Constant(value=base_currency), inner_toFloat],
+            )
```
Doesn't work Unable to resolve field: USD

Commit:
Fixing bot comments

Commit:
Fixing all sort of types

Commit:
Fixing conflicts

robbie-c reviewed (approved):
Ok, let's merge this, and the outstanding stuff can be added to the todo list

Commit:
migrating all to ast expressions

Commit:
Merge remote-tracking branch 'origin/master' into feat/conversion-goal-v2-ma

Commit:
Merge remote-tracking branch 'origin/master' into feat/conversion-goal-v2-ma
